wp-content/plugins/goonj-blocks/build/functions.php

```diff
@@ -15,4 +15,207 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
         throw new InvalidArgumentException('End time must be after start time');
     }
     return $start_date->format($format) . ' - ' . $end_date->format($format);
-}
\ No newline at end of file
+}
+/**
+ * Generate induction slots based on contact's location, registration date, and induction activity details.
+ * 
+ * @param int|null $contactId The ID of the contact for whom to generate slots.
+ * @param int $days The number of days to look ahead for slot generation (default is 30 days).
+ * @return array An array of induction slots or empty if criteria aren't met.
+ */
+function generate_induction_slots($contactId = null, $days = 30) {
+    // Validate contactId
+    if (empty($contactId) || !is_numeric($contactId)) {
+        \Civi::log()->warning('Invalid contactId', ['contactId' => $contactId]);
+        return [];
+    }
+
+    try {
+        // Fetch necessary contact details for slot generation
+        $contactData = \Civi\Api4\Contact::get(FALSE)
+            ->addSelect('address_primary.state_province_id', 'address_primary.city', 'Individual_fields.Created_Date')
+            ->addWhere('id', '=', $contactId)
+            ->execute()->single();
+
+        if (empty($contactData)) {
+            \Civi::log()->warning('Contact not found', ['contactId' => $contactId]);
+            return [];
+        }
+
+        // Retrieve induction activity for the contact
+        $inductionActivities = \Civi\Api4\Activity::get(FALSE)
+            ->addSelect('id', 'activity_date_time', 'status_id', 'status_id:name', 'Induction_Fields.Goonj_Office')
+            ->addWhere('source_contact_id', '=', $contactId)
+            ->addWhere('activity_type_id:name', '=', 'Induction')
+            ->execute();
+
+        // Return if no induction activity found
+        if ($inductionActivities->count() === 0) {
+            \Civi::log()->info('No activity found for contact', ['contactId' => $contactId]);
+            return [];
+        }
+
+        // If the induction activity status is 'Scheduled' or 'Completed', return the details
+        $inductionActivity = $inductionActivities->first();
```

Comment:
Can we use single on the API itself so we don't need to add first here

Reply:
sure